### PR TITLE
feat: disk-based shuffle for IVF partition assignment

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/lance/src/index/vector/ivf/shuffler.rs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                        vec![Arc::new(UInt32Array::from(vec![i]))],
+                    )
+                    .unwrap(),
+                )
+                .await
+                .unwrap();
+        }
+        let reader = shuffler.finish().await.unwrap();
+        assert_eq!(reader.keys(), vec![0, 1, 2]);
+        for i in 0..3 {
```
Comment (wjones127) [inline rust/lance/src/index/vector/ivf/shuffler.rs:188]: Could we add a test for what happens when you request group id = 3 (where there isn't any entry?)

File: rust/lance/src/io/writer.rs [hunk 2]
```diff
@@ -110,14 +110,18 @@ pub async fn write_manifest(
 pub struct FileWriter {
     object_writer: ObjectWriter,
     schema: Schema,
-    batch_id: i32,
+    pub batch_id: i32,
```
Comment (wjones127) [inline rust/lance/src/io/writer.rs:113]: Maybe this should be an accessor method? Also we should document what this is (is it the last batch we wrote, or the next one?)

File: rust/lance/src/index/vector/ivf/shuffler.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+use crate::{
+    error::Result,
+    io::{FileWriter, ObjectStore, RecordBatchStream},
+};
+
+const BUFFER_FILE_NAME: &str = "buffer.lance";
+
+/// Shuffle [RecordBatch] based on their IVF partition.
+///
+///
```
Comment (wjones127) [inline rust/lance/src/index/vector/ivf/shuffler.rs:35]: Could we add a bit more commentary here. I was thinking at first we wrote a file per partition, and then later realized it's all one file with the groups.

```suggestion
/// This buffers all the batches into a single Lance file, but keeps
/// track of which Lance file groups belong to each partition.
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
A few minor comments. Looks good otherwise.

File: rust/lance/src/index/vector/ivf/shuffler.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+use crate::datatypes::Schema;
+use crate::io::reader::batches_stream;
+use crate::io::FileReader;
+use crate::{
+    error::Result,
+    io::{FileWriter, ObjectStore, RecordBatchStream},
+};
+
+const BUFFER_FILE_NAME: &str = "buffer.lance";
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf/shuffler.rs:31]: This is just a regular lance file under `/tmp` on the worker. These are ephemeral / staging files.

[the remark this answers is not in the document]
Comment (chebbyChefNEQ) [reply rust/lance/src/index/vector/ivf/shuffler.rs:31]: sounds good, I was mostly concerned with having complex listing logic, say when we do garbage collection

[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf/shuffler.rs:31]: Got it. This file is deleted after Shuffler lifetime is gone.

File: rust/lance/src/index/vector/ivf/shuffler.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf/shuffler.rs:188]: Done

File: rust/lance/src/index/vector/ivf/shuffler.rs (hunk 3 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf/shuffler.rs:37]: Done

File: rust/lance/src/io/writer.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance/src/io/writer.rs]: Fixed.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**